File: go/starter/template/internal/get_payment_intent_quote.go
package internal

import (
	"context"
	"log"

	"connectrpc.com/connect"
	"github.com/t-0-network/provider-sdk/go/api/tzero/v1/common"
	"github.com/t-0-network/provider-sdk/go/api/tzero/v1/payment_intent"
	"github.com/t-0-network/provider-sdk/go/api/tzero/v1/payment_intent/payment_intentconnect"
)

// GetPaymentIntentQuote fetches indicative quotes for a pay-in currency/amount.
//
// Beneficiary Provider role — Step 3B.1.
// Use this to check available rates before creating a payment intent. The actual
// settlement rate is determined when the pay-in provider confirms funds received.
func GetPaymentIntentQuote(ctx context.Context, paymentIntentClient payment_intentconnect.PaymentIntentServiceClient) {
	quote, err := paymentIntentClient.GetQuote(ctx, connect.NewRequest(&payment_intent.GetQuoteRequest{
		Currency: "EUR",
		Amount:   &common.Decimal{Unscaled: 500, Exponent: 0}, // end-user pays 500 EUR
	}))
	if err != nil {
		log.Printf("Error getting payment intent quote: %s\n", err.Error())
		return
	}

	switch quote.Msg.Result.(type) {
	case *payment_intent.GetQuoteResponse_Success_:
		log.Printf("Got %d best pay-in quotes and %d total quotes\n",
			len(quote.Msg.GetSuccess().GetBestQuotes()),
			len(quote.Msg.GetSuccess().GetAllQuotes()),
		)
	case *payment_intent.GetQuoteResponse_QuoteNotFound_:
		log.Println("No pay-in quotes available for this currency/amount")
	default:
		log.Println("Unknown response type")
	}
}

File: java/starter/template/src/main/java/network/t0/provider/internal/GetPaymentIntentQuote.java
package network.t0.provider.internal;

import network.t0.sdk.proto.tzero.v1.common.Decimal;
import network.t0.sdk.proto.tzero.v1.payment_intent.GetQuoteRequest;
import network.t0.sdk.proto.tzero.v1.payment_intent.GetQuoteResponse;
import network.t0.sdk.proto.tzero.v1.payment_intent.PaymentIntentServiceGrpc;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Beneficiary Provider role — Step 3B.1.
 *
 * <p>Fetches indicative quotes for a pay-in currency/amount. Use this to check
 * available rates before creating a payment intent. The actual settlement rate is
 * determined when the pay-in provider confirms funds received.
 */
public class GetPaymentIntentQuote {

    private static final Logger log = LoggerFactory.getLogger(GetPaymentIntentQuote.class);

    public static void fetch(PaymentIntentServiceGrpc.PaymentIntentServiceBlockingStub paymentIntentClient) {
        try {
            GetQuoteResponse response = paymentIntentClient.getQuote(GetQuoteRequest.newBuilder()
                    .setCurrency("EUR")
                    .setAmount(Decimal.newBuilder().setUnscaled(500).setExponent(0).build()) // 500 EUR
                    .build());

            switch (response.getResultCase()) {
                case SUCCESS -> log.info("Got {} best pay-in quotes and {} total quotes",
                        response.getSuccess().getBestQuotesCount(),
                        response.getSuccess().getAllQuotesCount());
                case QUOTE_NOT_FOUND -> log.info("No pay-in quotes available for this currency/amount");
                default -> log.info("Unknown response type");
            }
        } catch (io.grpc.StatusRuntimeException e) {
            log.error("Error getting payment intent quote: {} - {}", e.getStatus().getCode(), e.getMessage());
        }
    }
}
